### ISSUE TITLE
A tutorial on step by step creation of IMDB for our own dataset and using pre-trained models for the same

### ISSUE THREAD

**Issue by raaju-shiv** [vlfeat/matconvnet#895]
It would be great if someone can come up with a step-by-step procedure of creating a imdb structure (say for three classes that supports multi-class classification) and using a pre-trained model to do the task. I could see newbies like me struggle in the very intial step of creating a imdb structure. Refering to CNN examples does not help understanding to beginners like us. If someone can come up with a tailored approach of the necessary modifications to be made to the mandatory functions (cnn_train, cnn_train_dag, cnn_cifar, cnn_cifar_init) for a custom owned dataset, I'm sure this gitHub issue will be an extremely useful practical for all those newbies like me. Thanks for your assistance.

**Comment by jotaf98** [comment]
You're right that even the MNIST example may have a bit too much for a simple example. I put together an example that is hopefully clear, and includes comments explaining it step by step:

http://www.robots.ox.ac.uk/~joao/cnn_toy_data.m

I hope to include this or something like it in the near future in the library.

**Comment by raaju-shiv** [comment]
Hi Henriques,

That is a great post on working with CNN. Many thanks for that. My principle concern in on the creation of the IMDB structure and I sincerely request you to help me with your modified imdb function alone, for this problem.

 I am working with medical images of Brain-tumors, where all my data are present in the root folder C:\Brain_tumors, in the form of sub-folders 'Brain-Tumor1', 'Brain-Tumor2', 'Brain-Tumor3',  'Brain-Tumor4', 'Brain-Tumor5'. Each sub-folder contains 5000 images of a specific type of tumor. All are RGB images of varying sizes. 
I need to create an imdb structure for this data. The following needs to be done:

1. All images need to be resized to 32*32*3.
2. The mean of the images in all sub-folders has to be removed
3. Contrast normalization and data whitening (similar to the imdbcifar example) need to be done for these 5000 images, in each of the sub-folders (i.e. there are 5 classes). This is mandatory as far as the medical images are concerned.
4. 30% of the images in these sub-folders need to be kept for training and the remaining 70% for validatation. No need for a separate test set for my validation becomes my test set.
5. Finally, we are going to have 4-D data, 5 labels and 2 sets.

I sincerely hope the modified imdb function (with the mention of how to point to the folders and sub-folders with the fullfile function), which you are going to help me with, will be a more comprhensive piece of code for everyone, working with medical images, attemtping to create an imdb structure, where we just need to change the number of classes and the number of images in a given class. 

Your assistance is very much needed. Thanks in advance.

**Comment by jotaf98** [comment]
You can mean-remove the images with
images = bsxfun(@minus, images, mean(images, 4)) ;
in getImdb.

The FOR loop in getImdb is where you load each of your images, and apply pre-processing as necessary.
What to do inside is very much application-dependent, and it's up to you to retrieve data from your folders and decide what Image Processing Toolbox functions to use to achieve any normalization/resizing/etc ;)

**Comment by samo313** [comment]
Hi 
I have the same question could you please upload his file again? it's not available
thanks

**Comment by jotaf98** [comment]
I moved it to the official examples directory, look here:
https://github.com/vlfeat/matconvnet/tree/master/examples/custom_imdb

I updated the original comment to prevent any future confusion ;)

**Comment by samo313** [comment]
Thanks a lot, I have one more question did you use matconvnet for binary
classification?

**Comment by jotaf98** [comment]
Yes, this example has 3 classes, just adapt it for 2 classes and you're ready to go!

**Comment by samo313** [comment]
Actually, I deal with an error in binary classification by using this code,
my labels are 0/1 and during test step I've got this error:
Error using vl_nnsoftmaxloss (line 52)
Assertion failed.
maybe some setting would be need for binary classification.
i set erorFunction to 'binary' and in last layer i edit 3 to 2 and i did
not any more change. would you please help me?

thanks in advance.


Error in vl_simplenn (line 332)
      res(i+1).x = vl_nnsoftmaxloss(res(i).x, l.class) ;


On Dec 23, 2017 7:05 PM, "Joao Henriques" <notifications@github.com> wrote:

Yes, this example has 3 classes, just adapt it for 2 classes and you're
ready to go!

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/vlfeat/matconvnet/issues/895#issuecomment-353732397>,
or mute the thread

**Comment by Tehseen-Akhtar** [comment]
you just need to change the last layer of you resulting network to softmax form softmaxloss and you are ready to go.
here is how it is done

ans.layers{end}.type = 'softmax';

enjoy

**Comment by Tehseen-Akhtar** [comment]
@jotaf98 
hi brother there is some logical type of error in this example that I am facing..
I trained your network and it converged very neatly..
when I tested it using the below code:

%First I read a random image from your training set..
im = imread(fullfile(vl_rootnn, 'data', 'toy-dataset/train/3/0130.png')) ;
%Converted it from logical to single..
im_ = single(im) ;
%forced the last layer from softmaxloss to softmax because it gives error otherwise
ans.layers{end}.type = 'softmax';
%loaded the image data base which contains mean of all the images..
load('imdb.mat');
subtracted the mean from the test image..
im_ = im_ - mean ;
%feeding the trained network and the test image to vl_simplenn for results
res = vl_simplenn(ans, im_) ;
%gathering the scores..
scores = squeeze(gather(res(end).x)) ;
[bestScore, best] = max(scores) ;
...
Now the problem with the above code is that whether I select an image from label 1, 2 or 3 the answer is always [0,1,0]
means the second category always wins
what am I doing wrong kindly elaborate 
thanks in advance..

Warm Regards
